base/linalg/triangular.jl

```diff
@@ -1867,48 +1867,53 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
 end
 logm(A::LowerTriangular) = logm(A.').'
 
-function sqrtm{T}(A::UpperTriangular{T})
-    n = checksquare(A)
+function sqrtm(A::UpperTriangular)
     realmatrix = false
     if isreal(A)
         realmatrix = true
-        for i = 1:n
+        for i = 1:checksquare(A)
             if real(A[i,i]) < 0
                 realmatrix = false
                 break
             end
         end
     end
-    if realmatrix
-        TT = typeof(sqrt(zero(T)))
-    else
-        TT = typeof(sqrt(complex(-one(T))))
-    end
-    R = zeros(TT, n, n)
-    for j = 1:n
-        R[j,j] = realmatrix?sqrt(A[j,j]):sqrt(complex(A[j,j]))
-        for i = j-1:-1:1
-            r = A[i,j]
-            for k = i+1:j-1
-                r -= R[i,k]*R[k,j]
+    sqrtm(A,Val{realmatrix})
+end
+function sqrtm{T,realmatrix}(A::UpperTriangular{T},::Type{Val{realmatrix}})
+    n = checksquare(A)
+    t = realmatrix ? typeof(sqrt(zero(T))) : typeof(sqrt(complex(zero(T))))
+    R = zeros(t, n, n)
+    tt = typeof(zero(t)*zero(t))
+    @inbounds begin
+        for j = 1:n
+            R[j,j] = realmatrix ? sqrt(A[j,j]) : sqrt(complex(A[j,j]))
+            for i = j-1:-1:1
+                r::tt = A[i,j]
+                @simd for k = i+1:j-1
+                    r -= R[i,k]*R[k,j]
+                end
+                r==0 || (R[i,j] = r / (R[i,i] + R[j,j]))
             end
-            r==0 || (R[i,j] = r / (R[i,i] + R[j,j]))
         end
     end
     return UpperTriangular(R)
 end
 function sqrtm{T}(A::UnitUpperTriangular{T})
     n = checksquare(A)
-    TT = typeof(sqrt(zero(T)))
-    R = zeros(TT, n, n)
-    for j = 1:n
-        R[j,j] = one(T)
-        for i = j-1:-1:1
-            r = A[i,j]
-            for k = i+1:j-1
-                r -= R[i,k]*R[k,j]
+    t = typeof(sqrt(zero(T)))
+    R = eye(t, n, n)
+    tt = typeof(zero(t)*zero(t))
+    half = 1/(2*R[1,1])
+    @inbounds begin
+        for j = 1:n
+            for i = j-1:-1:1
+                r::tt = A[i,j]
+                @simd for k = i+1:j-1
+                    r -= R[i,k]*R[k,j]
+                end
+                r==0 || (R[i,j] = half*r)
```

Comment:
See my comment above for the general case.  However, even for non-commutative algebras the case here is special because `R[i,i]` is the identity, which always commutes with everything.  So, the problem reduces to solving `R[i,j] + R[i,j] = r`.  *If* the elements form a field over the reals, then the solution `R[i,j] = r/2` is correct.

The case here is also wrong if e.g. the elements are matrices, in which case `1/(2*R[1,1])` will throw an error, although if you do `half = inv(2*R[1,1])` it will work.   However, it would be even better to just to `R[i,j] = r/2` in that case, since it would avoid the matrix multiplication.

However, it is still wrong if the elements are some other number field where you can't do `/2`. The correct, general solution should be
```julia
half=inv(R[1,1] + R[1,1])
```